sklearn/inspection/_permutation_importance.py

```diff
@@ -28,17 +28,19 @@ def _safe_column_indexing(X, col_idx):
 def _calculate_permutation_scores(estimator, X, y, col_idx, random_state,
                                   n_repeats, scorer):
     """Calculate score when `col_idx` is permuted."""
-    original_feature = _safe_column_indexing(X, col_idx).copy()
-    temp = original_feature.copy()
+    random_state = check_random_state(random_state)
 
+    # Work on a copy of X to to ensure thread-safety in case of threading
+    # based parallelism:
+    X_permuted = X.copy()
+    column_data = _safe_column_indexing(X_permuted, col_idx)
     scores = np.zeros(n_repeats)
     for n_round in range(n_repeats):
-        random_state.shuffle(temp)
-        _safe_column_setting(X, col_idx, temp)
-        feature_score = scorer(estimator, X, y)
+        random_state.shuffle(column_data)
+        _safe_column_setting(X_permuted, col_idx, column_data)
```

Comment:
This line might actually not be necessary as I assume that `_safe_column_indexing` always returns a view, or are their cases where it does not?

Reply:
This line is not needed, which means `_safe_column_setting` is also not needed. In this case `.values` returns a view in all cases because it always targets a column.

Reply:
not with a dataframe:

```python
In [25]: df.head()                                                                                                                                                                                    
Out[25]: 
          0         1         2         3         4         5         6         7         8         9
0  0.707264  0.538677 -2.530605  2.141858  0.413131  0.233331 -0.617173 -2.113002 -0.639021  0.893499
1 -2.648563 -0.242019 -2.107257  1.723822  0.930159  0.596130  0.110007 -0.653444 -0.522389 -0.404603
2  0.916806 -1.210205  0.296295 -0.876063 -0.006340  0.499527 -0.269714 -0.825826  0.569532 -1.053754
3 -0.721962 -0.718920  1.229692 -0.535799  0.297287  1.423598  0.659414 -0.020735 -0.462565  0.395459
4 -0.878993  0.673878  0.791935 -1.830175 -1.850800 -0.706001 -0.425039 -0.541195  0.426476  0.344111

In [26]: col = _safe_column_indexing(df, [0])                                                                                                                                                         

In [27]: col[0] = 10000                                                                                                                                                                               

In [28]: df.head()                                                                                                                                                                                    
Out[28]: 
          0         1         2         3         4         5         6         7         8         9
0  0.707264  0.538677 -2.530605  2.141858  0.413131  0.233331 -0.617173 -2.113002 -0.639021  0.893499
1 -2.648563 -0.242019 -2.107257  1.723822  0.930159  0.596130  0.110007 -0.653444 -0.522389 -0.404603
2  0.916806 -1.210205  0.296295 -0.876063 -0.006340  0.499527 -0.269714 -0.825826  0.569532 -1.053754
3 -0.721962 -0.718920  1.229692 -0.535799  0.297287  1.423598  0.659414 -0.020735 -0.462565  0.395459
4 -0.878993  0.673878  0.791935 -1.830175 -1.850800 -0.706001 -0.425039 -0.541195  0.426476  0.344111
```

Reply:
However, we don't need `_safe_column_indexing` I think. We can just use `_safe_indexing` with `axis=1`. `rng.shuffle(...)` will just be equivalent to calling `.values`

Reply:
After trying, this is a really bad idea to not convert to a numpy array. Not really sure what is the issue but this x10 slower. I should profile to know why :)

Reply:
> This line is not needed, which means _safe_column_setting is also not needed. In this case, .values returns a view in all cases because it always targets a column.

I should miss something, could you elaborate. It seems to not behave like a view. What am I missing?

Reply:
For this specific case, `col_idx` in `_safe_column_indexing` is always an integer:

```py
df = pd.DataFrame({'a': [1, 2, 3], 'b': [3, 4, 5]})
first = _safe_column_indexing(df, 0)
first[0] = 10
```

will update the dataframe.

Reply:
Given this discussion, we should add a comment to `_safe_column_indexing` if we make this change.

Reply:
> Given this discussion, we should add a comment to _safe_column_indexing if we make this change.

OK I did some fancy indexing there, that's the reason. Thanks @thomasjpfan

Reply:
what about being explicit and not writing a function since we only call it once